sui_programmability/examples/games/sources/frenemies.move

```diff
@@ -0,0 +1,387 @@
+        id: UID,
+        /// current assignment
+        assignment: Assignment,
+        /// accumulated score across epochs
+        score: u16,
+        /// number of epochs for which the player received a score (even if the score was 0)
+        participation: u16,
+        /// last epoch for which a score has been recorded
+        epoch: u64
+    }
+
+    /// Information about a validator
+    struct Validator has store, drop, copy {
+        addr: address,
+        stake: u64,
+    }
+
+    /// An epoch assignment for a player in the game
+    struct Assignment has store, drop {
+        /// validator this assignment is directed at
+        validator: address,
+        /// goal of the assignment: make validator one of {friend, enemy, neutral} at end of epoch
+        goal: u8,
+    }
+
+    /// Event emitted each time a player's score is updated. Suitable for building a scoreboard offchain
+    struct NewScoreEvent has copy, drop {
+        /// Address of the player
+        addr: address,
+        /// Object ID of the player's scorecard. Note: a single address can have multiple cards, but they'll all have the same assignment
+        scorecard: ID,
+        /// Total score
+        score: u16,
+        /// Number of epochs the player has scored in
+        participation: u16,
+        /// Epoch when this score was recorded
+        epoch: u64
+    }
+
+    /// Goal: validator finishes in top third by stake
+    const FRIEND: u8 = 0;
+    /// Goal: validator finishes in middle third by stake
+    const NEUTRAL: u8 = 1;
+    /// Goal: validator finishes in bottom third by stake
+    const ENEMY: u8 = 2;
+
+    /// Number of points a player receives for successfully completing an epoch goal.
+    /// Additional points are awarded according to the difficulty of the goal (i.e.,
+    /// how many places the validator moved during the epoch)
+    const GOAL_COMPLETION_POINTS: u16 = 10;
+
+    /// Attempting to call `score` on the same `Scorecard` object more than once/epoch
+    const EAlreadyUpdatedScorecardThisEpoch: u64 = 0;
+
+    fun init(ctx: &mut TxContext) {
+        let epoch = tx_context::epoch(ctx);
+        transfer::share_object(
+            Leaderboard {
+                id: object::new(ctx),
+                prev_epoch_stakes: table::new(ctx),
+                epoch,
+                start_epoch: epoch,
+            }
+        )
+    }
+
+    /// Get the current assignment for `addr`.
+    /// Always returns the same assignment for a fixed address A and epoch E.
+    fun get_assignment(addr: address, state: &SuiSystemState): Assignment {
+        let validators = validator_set::active_validators(sui_system::validators(state));
+        let len = vector::length(validators);
+
+        // derive assignment from the stake of top validator + address.
+        // this means that for a fixed address A and epoch E, the assignment is deterministic
+        let top_stake = (validator::total_stake(vector::borrow(validators, 0)) as u256);
```

Comment:
Right now we just take the address of the first validator in the active validator vector, which is unsorted, so I wouldn't call it "top_stake".

Reply:
Good point + will fix.